File: backend/app/services/image_upload_service.py
"""Backblaze B2 uploads via the image-upload Cloudflare Worker."""

from __future__ import annotations

import uuid
from pathlib import Path
from urllib.parse import urlparse

import requests
from flask import current_app


class ImageUploadError(Exception):
    def __init__(self, message: str, status_code: int | None = None):
        super().__init__(message)
        self.status_code = status_code


_MIME_BY_EXT = {
    "pdf": "application/pdf",
    "png": "image/png",
    "jpg": "image/jpeg",
    "jpeg": "image/jpeg",
    "webp": "image/webp",
}

_EXT_BY_MIME = {
    "application/pdf": "pdf",
    "image/png": "png",
    "image/jpeg": "jpg",
    "image/webp": "webp",
}

_ALLOWED_LOCAL_PREFIXES = ("packages/", "invoices/", "transfer-proofs/")


def guess_content_type(filename: str, default: str = "application/octet-stream") -> str:
    ext = filename.rsplit(".", 1)[-1].lower() if "." in filename else ""
    return _MIME_BY_EXT.get(ext, default)


def parse_content_length(data: dict) -> int:
    raw = data.get("content_length")
    if raw is None:
        raw = data.get("contentLength")
    if raw is None:
        raw = data.get("size")
    if raw is None:
        raise ValueError("content_length is required")
    try:
        length = int(raw)
    except (TypeError, ValueError) as exc:
        raise ValueError("content_length is required") from exc
    if length <= 0:
        raise ValueError("content_length must be greater than 0")
    return length


def parse_presign_fields(
    data: dict,
    *,
    default_filename: str,
    default_content_type: str,
) -> tuple[str, str, int]:
    filename = (data.get("filename") or default_filename).strip()
    content_type = (
        data.get("content_type") or data.get("contentType") or ""
    ).strip().lower()
    if not content_type:
        content_type = guess_content_type(filename, default_content_type)
    content_length = parse_content_length(data)
    return filename, content_type, content_length


def worker_base_url() -> str:
    base = (current_app.config.get("IMAGE_UPLOAD_WORKER_URL") or "").strip().rstrip("/")
    if base.endswith("/upload-worker"):
        base = base[: -len("/upload-worker")]
    if base:
        return base
    full = (current_app.config.get("IMAGE_UPLOAD_URL") or "").strip().rstrip("/")
    if full.endswith("/upload-url"):
        return full[: -len("/upload-url")]
    if full.endswith("/upload-worker"):
        return full[: -len("/upload-worker")]
    return full


def worker_api_key() -> str:
    return (
        current_app.config.get("IMAGE_UPLOAD_API_KEY")
        or current_app.config.get("IMAGE_API_KEY")
        or ""
    )


def is_image_upload_configured() -> bool:
    return bool(worker_base_url() and worker_api_key())


def is_local_upload_enabled() -> bool:
    if is_image_upload_configured():
        return False
    return bool(current_app.config.get("LOCAL_UPLOADS_ENABLED", True))


def local_upload_root() -> Path:
    configured = (current_app.config.get("LOCAL_UPLOAD_ROOT") or "").strip()
    if configured:
        return Path(configured)
    return Path(current_app.root_path).parent / "var" / "uploads"


def is_storage_configured() -> bool:
    return is_image_upload_configured() or is_local_upload_enabled()


def is_allowed_local_object_key(object_key: str) -> bool:
    normalized = object_key.replace("\\", "/").lstrip("/")
    if not normalized or ".." in normalized.split("/"):
        return False
    return any(normalized.startswith(prefix) for prefix in _ALLOWED_LOCAL_PREFIXES)


def local_public_url(object_key: str) -> str:
    return f"/api/uploads/files/{object_key.lstrip('/')}"


def local_file_path(object_key: str) -> Path:
    if not is_allowed_local_object_key(object_key):
        raise ValueError("Invalid upload object key")
    return local_upload_root() / object_key.replace("\\", "/").lstrip("/")


def save_local_upload(object_key: str, file_bytes: bytes) -> None:
    path = local_file_path(object_key)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(file_bytes)


def create_local_presigned_upload(
    *,
    content_type: str,
    content_length: int,
    prefix: str,
) -> dict:
    if content_length <= 0:
        raise ImageUploadError("content_length must be positive")

    ext = _EXT_BY_MIME.get(content_type, "bin")
    object_key = f"{prefix.strip('/')}/{uuid.uuid4()}.{ext}"
    if not is_allowed_local_object_key(object_key):
        raise ImageUploadError("Invalid upload prefix")

    return {
        "upload_url": "local",
        "upload_headers": {},
        "public_url": local_public_url(object_key),
        "object_key": object_key,
    }


def key_from_public_url(public_url: str) -> str:
    return urlparse(public_url).path.lstrip("/")


def resolve_stored_url(stored: str | None) -> str | None:
    """Return a browser-loadable URL for a stored object key or full URL."""
    if not stored:
        return None
    if stored.startswith(("http://", "https://")):
        return stored
    public_base = (current_app.config.get("STORAGE_PUBLIC_URL") or "").strip().rstrip("/")
    if public_base:
        return f"{public_base}/{stored.lstrip('/')}"
    if is_local_upload_enabled() and is_allowed_local_object_key(stored):
        try:
            if local_file_path(stored).is_file():
                return local_public_url(stored)
        except ValueError:
            return None
    return None


def create_upload_presign(
    *,
    content_type: str,
    content_length: int,
    prefix: str,
) -> dict:
    """Presigned PUT URL from the image upload worker."""
    return create_presigned_upload(
        content_type=content_type,
        content_length=content_length,
        prefix=prefix,
    )


def create_presigned_upload(
    *,
    content_type: str,
    content_length: int,
    prefix: str = "packages",
) -> dict:
    """
    Step A of the upload flow — presigned PUT URL from the Worker.

    Returns upload_url, upload_headers, public_url, object_key.
    """
    base = worker_base_url()
    api_key = worker_api_key()
    if not base or not api_key:
        if is_local_upload_enabled():
            return create_local_presigned_upload(
                content_type=content_type,
                content_length=content_length,
                prefix=prefix,
            )
        raise ImageUploadError("Image upload worker is not configured")

    if content_length <= 0:
        raise ImageUploadError("content_length must be positive")

    try:
        response = requests.post(
            f"{base}/upload-url",
            json={
                "contentType": content_type,
                "contentLength": content_length,
                "prefix": prefix,
            },
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}",
            },
            timeout=30,
        )
    except requests.RequestException as exc:
        raise ImageUploadError(f"Presign request failed: {exc}") from exc

    try:
        data = response.json()
    except ValueError:
        raise ImageUploadError(
            response.text or "Invalid presign response",
            status_code=response.status_code,
        )

    if not response.ok:
        raise ImageUploadError(
            data.get("error", "Presign failed"),
            status_code=response.status_code,
        )

    public_url = data.get("publicUrl") or data.get("public_url")
    upload_url = data.get("uploadUrl") or data.get("upload_url")
    if not public_url or not upload_url:
        raise ImageUploadError("Presign response missing uploadUrl or publicUrl")

    object_key = (data.get("key") or data.get("object_key") or "").strip()
    if not object_key:
        object_key = key_from_public_url(public_url)

    return {
        "upload_url": upload_url,
        "upload_headers": data.get("headers") or data.get("upload_headers") or {},
        "public_url": public_url,
        "object_key": object_key,
    }


def complete_presigned_upload(
    *,
    upload_url: str,
    file_bytes: bytes,
    content_type: str,
    upload_headers: dict | None = None,
) -> None:
    headers = {**(upload_headers or {}), "Content-Type": content_type}
    try:
        upload = requests.put(
            upload_url,
            data=file_bytes,
            headers=headers,
            timeout=120,
        )
        upload.raise_for_status()
    except requests.RequestException as exc:
        raise ImageUploadError(f"Upload failed: {exc}") from exc


def upload_bytes(file_bytes: bytes, content_type: str, *, prefix: str = "packages") -> str:
    """Server-side upload (Step A + B). Returns publicUrl."""
    presign = create_presigned_upload(
        content_type=content_type,
        content_length=len(file_bytes),
        prefix=prefix,
    )
    complete_presigned_upload(
        upload_url=presign["upload_url"],
        file_bytes=file_bytes,
        content_type=content_type,
        upload_headers=presign["upload_headers"],
    )
    return presign["public_url"]


def is_valid_photo_reference(key: str, *, unidentified: bool = False, shipping_id: str | None = None) -> bool:
    if not key:
        return False
    if key.startswith(("http://", "https://")):
        return True
    if unidentified:
        return key.startswith("packages/unidentified/") or key.startswith("packages/")
    if shipping_id and key.startswith(f"packages/{shipping_id}/"):
        return True
    return key.startswith("packages/")


def is_valid_invoice_reference(key: str, shipping_id: str | None = None) -> bool:
    if not key:
        return False
    if key.startswith(("http://", "https://")):
        return True
    if shipping_id and key.startswith(f"invoices/{shipping_id}/"):
        return True
    return key.startswith("invoices/")


def is_valid_transfer_proof_reference(key: str) -> bool:
    if not key:
        return False
    if key.startswith(("http://", "https://")):
        return True
    return key.startswith("transfer-proofs/")

File: server/src/services/imageUploadService.ts
import { existsSync } from 'node:fs'
import { mkdir, writeFile } from 'node:fs/promises'
import { dirname, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import { randomUUID } from 'node:crypto'
import { config } from '../config.js'

export class ImageUploadError extends Error {
  constructor(
    message: string,
    readonly statusCode: number | null = null,
  ) {
    super(message)
    this.name = 'ImageUploadError'
  }
}

const MIME_BY_EXT: Record<string, string> = {
  pdf: 'application/pdf',
  png: 'image/png',
  jpg: 'image/jpeg',
  jpeg: 'image/jpeg',
  webp: 'image/webp',
}

const EXT_BY_MIME: Record<string, string> = {
  'application/pdf': 'pdf',
  'image/png': 'png',
  'image/jpeg': 'jpg',
  'image/webp': 'webp',
}

const ALLOWED_LOCAL_PREFIXES = ['packages/', 'invoices/', 'transfer-proofs/']

export function guessContentType(filename: string, defaultType = 'application/octet-stream'): string {
  const ext = filename.includes('.') ? filename.split('.').pop()!.toLowerCase() : ''
  return MIME_BY_EXT[ext] ?? defaultType
}

export function parseContentLength(data: Record<string, unknown>): number {
  const raw = data.content_length ?? data.contentLength ?? data.size
  if (raw == null) throw new Error('content_length is required')
  const length = parseInt(String(raw), 10)
  if (Number.isNaN(length)) throw new Error('content_length is required')
  if (length <= 0) throw new Error('content_length must be greater than 0')
  return length
}

export function parsePresignFields(
  data: Record<string, unknown>,
  opts: { defaultFilename: string; defaultContentType: string },
): [string, string, number] {
  const filename = String(data.filename ?? opts.defaultFilename).trim()
  let contentType = String(data.content_type ?? data.contentType ?? '')
    .trim()
    .toLowerCase()
  if (!contentType) contentType = guessContentType(filename, opts.defaultContentType)
  const contentLength = parseContentLength(data)
  return [filename, contentType, contentLength]
}

export function workerBaseUrl(): string {
  let base = config.imageUploadWorkerUrl.trim().replace(/\/$/, '')
  if (base.endsWith('/upload-worker')) base = base.slice(0, -'/upload-worker'.length)
  if (base) return base
  let full = config.imageUploadUrl.trim().replace(/\/$/, '')
  if (full.endsWith('/upload-url')) full = full.slice(0, -'/upload-url'.length)
  if (full.endsWith('/upload-worker')) full = full.slice(0, -'/upload-worker'.length)
  return full
}

export function workerApiKey(): string {
  return config.imageUploadApiKey || config.imageApiKey || ''
}

export function isImageUploadConfigured(): boolean {
  return Boolean(workerBaseUrl() && workerApiKey())
}

export function isLocalUploadEnabled(): boolean {
  if (isImageUploadConfigured()) return false
  return config.localUploadsEnabled
}

export function localUploadRoot(): string {
  if (config.localUploadRoot.trim()) return config.localUploadRoot.trim()
  const serverRoot = resolve(fileURLToPath(new URL('../..', import.meta.url)))
  return resolve(serverRoot, '../var/uploads')
}

export function isStorageConfigured(): boolean {
  return isImageUploadConfigured() || isLocalUploadEnabled()
}

export function isAllowedLocalObjectKey(objectKey: string): boolean {
  const normalized = objectKey.replace(/\\/g, '/').replace(/^\//, '')
  if (!normalized || normalized.split('/').includes('..')) return false
  return ALLOWED_LOCAL_PREFIXES.some((prefix) => normalized.startsWith(prefix))
}

export function localPublicUrl(objectKey: string): string {
  return `/api/uploads/files/${objectKey.replace(/^\//, '')}`
}

export function localFilePath(objectKey: string): string {
  if (!isAllowedLocalObjectKey(objectKey)) throw new Error('Invalid upload object key')
  return resolve(localUploadRoot(), objectKey.replace(/\\/g, '/').replace(/^\//, ''))
}

export async function saveLocalUpload(objectKey: string, fileBytes: Buffer): Promise<void> {
  const path = localFilePath(objectKey)
  await mkdir(dirname(path), { recursive: true })
  await writeFile(path, fileBytes)
}

export function createLocalPresignedUpload(opts: {
  contentType: string
  contentLength: number
  prefix: string
}): Record<string, unknown> {
  if (opts.contentLength <= 0) throw new ImageUploadError('content_length must be positive')
  const ext = EXT_BY_MIME[opts.contentType] ?? 'bin'
  const objectKey = `${opts.prefix.replace(/\/$/, '')}/${randomUUID()}.${ext}`
  if (!isAllowedLocalObjectKey(objectKey)) throw new ImageUploadError('Invalid upload prefix')
  return {
    upload_url: 'local',
    upload_headers: {},
    public_url: localPublicUrl(objectKey),
    object_key: objectKey,
  }
}

export function keyFromPublicUrl(publicUrl: string): string {
  try {
    return new URL(publicUrl, 'http://local').pathname.replace(/^\//, '')
  } catch {
    return publicUrl.replace(/^\//, '')
  }
}

export function resolveStoredUrl(stored: string | null | undefined): string | null {
  if (!stored) return null
  if (stored.startsWith('http://') || stored.startsWith('https://')) return stored
  const publicBase = config.storagePublicUrl.trim().replace(/\/$/, '')
  if (publicBase) return `${publicBase}/${stored.replace(/^\//, '')}`
  if (isLocalUploadEnabled() && isAllowedLocalObjectKey(stored)) {
    try {
      if (existsSync(localFilePath(stored))) return localPublicUrl(stored)
    } catch {
      return null
    }
  }
  return null
}

export async function createPresignedUpload(opts: {
  contentType: string
  contentLength: number
  prefix?: string
}): Promise<Record<string, unknown>> {
  const prefix = opts.prefix ?? 'packages'
  const base = workerBaseUrl()
  const apiKey = workerApiKey()
  if (!base || !apiKey) {
    if (isLocalUploadEnabled()) {
      return createLocalPresignedUpload({ ...opts, prefix })
    }
    throw new ImageUploadError('Image upload worker is not configured')
  }
  if (opts.contentLength <= 0) throw new ImageUploadError('content_length must be positive')

  let response: Response
  try {
    response = await fetch(`${base}/upload-url`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({
        contentType: opts.contentType,
        contentLength: opts.contentLength,
        prefix,
      }),
    })
  } catch (err) {
    throw new ImageUploadError(`Presign request failed: ${err}`)
  }

  let data: Record<string, unknown>
  try {
    data = (await response.json()) as Record<string, unknown>
  } catch {
    throw new ImageUploadError((await response.text()) || 'Invalid presign response', response.status)
  }

  if (!response.ok) {
    throw new ImageUploadError(String(data.error ?? 'Presign failed'), response.status)
  }

  const publicUrl = (data.publicUrl ?? data.public_url) as string | undefined
  const uploadUrl = (data.uploadUrl ?? data.upload_url) as string | undefined
  if (!publicUrl || !uploadUrl) throw new ImageUploadError('Presign response missing uploadUrl or publicUrl')

  let objectKey = String(data.key ?? data.object_key ?? '').trim()
  if (!objectKey) objectKey = keyFromPublicUrl(publicUrl)

  return {
    upload_url: uploadUrl,
    upload_headers: data.headers ?? data.upload_headers ?? {},
    public_url: publicUrl,
    object_key: objectKey,
  }
}

export function createUploadPresign(opts: {
  contentType: string
  contentLength: number
  prefix: string
}): Promise<Record<string, unknown>> {
  return createPresignedUpload(opts)
}

export async function completePresignedUpload(opts: {
  uploadUrl: string
  fileBytes: Buffer
  contentType: string
  uploadHeaders?: Record<string, string>
}): Promise<void> {
  const headers = { ...(opts.uploadHeaders ?? {}), 'Content-Type': opts.contentType }
  try {
    const upload = await fetch(opts.uploadUrl, {
      method: 'PUT',
      body: new Uint8Array(opts.fileBytes),
      headers,
    })
    if (!upload.ok) throw new Error(`HTTP ${upload.status}`)
  } catch (err) {
    throw new ImageUploadError(`Upload failed: ${err}`)
  }
}

export function isValidInvoiceReference(key: string, shippingId?: string | null): boolean {
  if (!key) return false
  if (key.startsWith('http://') || key.startsWith('https://')) return true
  if (shippingId && key.startsWith(`invoices/${shippingId}/`)) return true
  return key.startsWith('invoices/')
}

export function isValidTransferProofReference(key: string): boolean {
  if (!key) return false
  if (key.startsWith('http://') || key.startsWith('https://')) return true
  return key.startsWith('transfer-proofs/')
}

export function isValidPhotoReference(
  key: string,
  opts: { unidentified?: boolean; shippingId?: string | null } = {},
): boolean {
  if (!key) return false
  if (key.startsWith('http://') || key.startsWith('https://')) return true
  if (opts.unidentified) {
    return key.startsWith('packages/unidentified/') || key.startsWith('packages/')
  }
  if (opts.shippingId && key.startsWith(`packages/${opts.shippingId}/`)) return true
  return key.startsWith('packages/')
}
